Using add group + after removing group causes a crash

#### Software and hardware versions 
Collect v1.28 current master, Android v5, 7, 10 probably all, device used...

#### Problem description
Right after deleting group and using + in the hierarchy to add another one Collect is freezing and crashing. I was able to reproduce it for both repeat groups and nested groups.

#### Steps to reproduce the problem
1. Use any form with repeat groups, I have used: 
[nested-repeats.xml.txt](https://github.com/getodk/collect/files/4918782/nested-repeats.xml.txt) 
2. Add some groups
3. Delete the last one
4. Click on + in the hierarchy (I can see that gif is lagging a bit but I definitely clicked on + button)

gif:
![ezgif com-video-to-gif (9)](https://user-images.githubusercontent.com/33342488/87425572-2d33c300-c5de-11ea-890f-6da77436e2b4.gif)

#### Expected behavior
Collect should not crash. Users should be able to add another group. 

#### Other information 
stacktrace:
`020-07-14 14:08:49.604 1970-2024/? E/ActivityManager: ANR in org.odk.collect.android (org.odk.collect.android/.activities.FormHierarchyActivity)
    PID: 30275
    Reason: Input dispatching timed out (Waiting to send non-key event because the touched window has not finished processing certain input events that were delivered to it over 500.0ms ago.  Wait queue length: 47.  Wait queue head age: 5514.1ms.)
    Load: 5.13 / 5.07 / 4.82
    CPU usage from 69312ms to 0ms ago (2020-07-14 14:07:37.544 to 2020-07-14 14:08:46.856):
      100% 30275/org.odk.collect.android: 100% user + 0% kernel / faults: 3 minor
      0.6% 1970/system_server: 0.3% user + 0.2% kernel / faults: 277 minor
      0.4% 29796/kworker/u8:1: 0% user + 0.4% kernel
      0.2% 21112/kworker/u8:4: 0% user + 0.2% kernel
      0.1% 30484/kworker/0:0: 0% user + 0.1% kernel
      0% 5580/com.motorola.motocare: 0% user + 0% kernel / faults: 223 minor
      0% 6/kworker/u8:0: 0% user + 0% kernel
      0% 347/surfaceflinger: 0% user + 0% kernel / faults: 13 minor
      0% 185/mmcqd/0: 0% user + 0% kernel
      0% 2453/wpa_supplicant: 0% user + 0% kernel / faults: 212 minor
      0% 256/logd: 0% user + 0% kernel / faults: 2 minor
      0% 2449/VosMCThread: 0% user + 0% kernel
      0% 2483/com.android.systemui: 0% user + 0% kernel / faults: 31 minor
      0% 5894/uk.org.invisibility.recordablefree: 0% user + 0% kernel / faults: 1 major
      0% 24791/com.google.process.gapps: 0% user + 0% kernel / faults: 203 minor
      0% 29824/irq/326-synapti: 0% user + 0% kernel
      0% 3/ksoftirqd/0: 0% user + 0% kernel
      0% 7/rcu_preempt: 0% user + 0% kernel
      0% 33/kworker/u9:0: 0% user + 0% kernel
      0% 333/healthd: 0% user + 0% kernel
      0% 345/lmkd: 0% user + 0% kernel
      0% 2450/VosTXThread: 0% user + 0% kernel
      0% 2736/com.android.phone: 0% user + 0% kernel / faults: 4 minor
      0% 24687/kworker/2:1: 0% user + 0% kernel
      0% 26325/kworker/0:2: 0% user + 0% kernel
      0% 28802/com.google.android.gms.persistent: 0% user + 0% kernel / faults: 1 minor
    25% TOTAL: 25% user + 0.3% kernel + 0% iowait + 0% softirq
    CPU usage from 2201ms to 2724ms later (2020-07-14 14:08:49.057 to 2020-07-14 14:08:49.579):
      99% 30275/org.odk.collect.android: 99% user + 0% kernel
        101% 30275/collect.android: 101% user + 0% kernel
      5.7% 1970/system_server: 1.9% user + 3.8% kernel
        3.8% 2024/ActivityManager: 1.9% user + 1.9% kernel
        1.9% 2398/SensorService: 0% user + 1.9% kernel
    26% TOTAL: 25% user + 0.4% kernel
2020-07-14 14:09:08.766 378-2243/? E/NetlinkEvent: NetlinkEvent::FindParam(): Parameter 'UID' not found
2020-07-14 14:09:11.802 1970-2403/? E/InputDispatcher: channel '63e84d8 org.odk.collect.android/org.odk.collect.android.activities.FormEntryActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
2020-07-14 14:09:12.488 30509-30570/org.odk.collect.android E/FirebaseInstanceId: Failed to get FIS auth token
    java.util.concurrent.ExecutionException: com.google.firebase.installations.FirebaseInstallationsException
        at com.google.android.gms.tasks.Tasks.zzb(Unknown Source)
        at com.google.android.gms.tasks.Tasks.await(Unknown Source)
        at com.google.firebase.iid.zzt.zzb(com.google.firebase:firebase-iid@@20.1.5:54)
        at com.google.firebase.iid.zzt.zza(com.google.firebase:firebase-iid@@20.1.5:72)
        at com.google.firebase.iid.zzs.run(com.google.firebase:firebase-iid@@20.1.5)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at com.google.android.gms.common.util.concurrent.zza.run(com.google.android.gms:play-services-basement@@17.1.1:6)
        at java.lang.Thread.run(Thread.java:761)
     Caused by: com.google.firebase.installations.FirebaseInstallationsException
        at com.google.firebase.installations.FirebaseInstallations.doNetworkCall(com.google.firebase:firebase-installations@@16.2.1:350)
        at com.google.firebase.installations.FirebaseInstallations.lambda$doRegistrationInternal$0(com.google.firebase:firebase-installations@@16.2.1:323)
        at com.google.firebase.installations.FirebaseInstallations$$Lambda$5.run(com.google.firebase:firebase-installations@@16.2.1)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761) 
2020-07-14 14:09:12.973 30509-30509/org.odk.collect.android E/PropertyManager: java.lang.SecurityException: getDeviceId: Neither user 10862 nor current process has android.permission.READ_PHONE_STATE.
        at android.os.Parcel.readException(Parcel.java:1684)
        at android.os.Parcel.readException(Parcel.java:1637)
        at com.android.internal.telephony.ITelephony$Stub$Proxy.getDeviceId(ITelephony.java:4726)
        at android.telephony.TelephonyManager.getDeviceId(TelephonyManager.java:866)
        at org.odk.collect.android.injection.config.AppDependencyModule$1.getDeviceId(AppDependencyModule.java:219)
        at org.odk.collect.android.logic.PropertyManager.findDeviceId(PropertyManager.java:125)
        at org.odk.collect.android.logic.PropertyManager.reload(PropertyManager.java:104)
        at org.odk.collect.android.logic.PropertyManager.<init>(PropertyManager.java:98)
        at org.odk.collect.android.application.Collect.initializeJavaRosa(Collect.java:174)
        at org.odk.collect.android.application.Collect.onCreate(Collect.java:130)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1046)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5444)
        at android.app.ActivityThread.-wrap2(ActivityThread.java)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1558)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6165)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:888)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:778)
2020-07-14 14:09:13.707 30509-30520/org.odk.collect.android E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
    java.lang.Throwable: Explicit termination method 'end' not called
        at dalvik.system.CloseGuard.open(CloseGuard.java:180)
        at java.util.zip.Inflater.<init>(Inflater.java:104)
        at com.android.okhttp.okio.GzipSource.<init>(GzipSource.java:62)
        at com.android.okhttp.internal.http.HttpEngine.unzip(HttpEngine.java:645)
        at com.android.okhttp.internal.http.HttpEngine.readResponse(HttpEngine.java:821)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:463)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:405)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:521)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode(DelegatingHttpsURLConnection.java:105)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java)
        at com.google.firebase.installations.remote.FirebaseInstallationServiceClient.createFirebaseInstallation(com.google.firebase:firebase-installations@@16.2.1:156)
        at com.google.firebase.installations.FirebaseInstallations.registerFidWithServer(com.google.firebase:firebase-installations@@16.2.1:433)
        at com.google.firebase.installations.FirebaseInstallations.doNetworkCall(com.google.firebase:firebase-installations@@16.2.1:333)
        at com.google.firebase.installations.FirebaseInstallations.lambda$doRegistrationInternal$0(com.google.firebase:firebase-installations@@16.2.1:323)
        at com.google.firebase.installations.FirebaseInstallations$$Lambda$5.run(com.google.firebase:firebase-installations@@16.2.1)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
2020-07-14 14:09:14.162 30509-30566/org.odk.collect.android E/FirebaseInstanceId: Topic sync or token retrieval failed on hard failure exceptions: FIS_AUTH_ERROR. Won't retry the operation.
2020-07-14 14:09:18.708 2736-2736/? E/QtiImsExtUtils: getConfigForPhoneId subId is invalid
2020-07-14 14:09:18.709 2736-2736/? E/QtiImsExtUtils: isCarrierConfigEnabled bundle is null
2020-07-14 14:09:19.487 30509-30520/org.odk.collect.android E/StrictMode: A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
    java.lang.Throwable: Explicit termination method 'end' not called
        at dalvik.system.CloseGuard.open(CloseGuard.java:180)
        at java.util.zip.Inflater.<init>(Inflater.java:104)
        at com.android.okhttp.okio.GzipSource.<init>(GzipSource.java:62)
        at com.android.okhttp.internal.http.HttpEngine.unzip(HttpEngine.java:645)
        at com.android.okhttp.internal.http.HttpEngine.readResponse(HttpEngine.java:821)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:463)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:405)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:521)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode(DelegatingHttpsURLConnection.java:105)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java)
        at com.google.firebase.installations.remote.FirebaseInstallationServiceClient.createFirebaseInstallation(com.google.firebase:firebase-installations@@16.2.1:156)
        at com.google.firebase.installations.FirebaseInstallations.registerFidWithServer(com.google.firebase:firebase-installations@@16.2.1:433)
        at com.google.firebase.installations.FirebaseInstallations.doNetworkCall(com.google.firebase:firebase-installations@@16.2.1:333)
        at com.google.firebase.installations.FirebaseInstallations.lambda$doRegistrationInternal$0(com.google.firebase:firebase-installations@@16.2.1:323)
        at com.google.firebase.installations.FirebaseInstallations$$Lambda$5.run(com.google.firebase:firebase-installations@@16.2.1)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
`

--- a/collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/AddRepeatTest.java
+++ b/collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/AddRepeatTest.java
@@ -122,7 +122,7 @@ public class AddRepeatTest {
                 .clickOnAdd(new FormEntryPage("One Question Repeat", rule))
                 .clickGoToArrow()
                 .clickGoUpIcon()
-                .clickPlus("Person")
+                .addGroup()
                 .assertText("Person > 3");
     }
 }

--- a/collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormHierarchyTest.java
+++ b/collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/FormHierarchyTest.java
@@ -37,6 +37,7 @@ public class FormHierarchyTest {
             .around(new CopyFormRule("formHierarchy1.xml", null))
             .around(new CopyFormRule("formHierarchy2.xml", null))
             .around(new CopyFormRule("formHierarchy3.xml", null))
+            .around(new CopyFormRule("repeat_group_new.xml", null))
             .around(rule);
 
     @Test
@@ -159,5 +160,22 @@ public class FormHierarchyTest {
         onView(withId(R.id.list)).check(matches(RecyclerViewMatcher.withListSize(1)));
 
         page.assertText("Repeat Group 1_1 > 1");
+    }
+
+    @Test
+    //https://github.com/getodk/collect/issues/3971
+    public void deletingLastGroupAndAddingOneShouldNotBreakHierarchy() {
+        new MainMenuPage(rule)
+                .startBlankFormWithRepeatGroup("RepeatGroupNew", "People")
+                .clickOnAdd(new FormEntryPage("RepeatGroupNew", rule))
+                .swipeToNextQuestion()
+                .swipeToNextQuestion()
+                .clickOnAddGroup()
+                .swipeToNextQuestion()
+                .swipeToNextQuestion()
+                .clickOnAddGroup()
+                .clickGoToArrow()
+                .deleteGroup()
+                .addGroup();
     }
 }

--- a/collect_app/src/androidTest/java/org/odk/collect/android/support/pages/FormHierarchyPage.java
+++ b/collect_app/src/androidTest/java/org/odk/collect/android/support/pages/FormHierarchyPage.java
@@ -29,13 +29,13 @@ public class FormHierarchyPage extends Page<FormHierarchyPage> {
         return this;
     }
 
-    public FormEntryPage clickPlus(String repeatName) {
-        onView(withId(R.id.menu_add_repeat)).perform(click());
+    public FormEntryPage clickGoToStart() {
+        onView(withId(R.id.jumpBeginningButton)).perform(click());
         return new FormEntryPage(formName, rule).assertOnPage();
     }
 
-    public FormEntryPage clickGoToStart() {
-        onView(withId(R.id.jumpBeginningButton)).perform(click());
+    public FormEntryPage addGroup() {
+        onView(withId(R.id.menu_add_repeat)).perform(click());
         return new FormEntryPage(formName, rule).assertOnPage();
     }
 

--- a/collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java
+++ b/collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java
@@ -320,6 +320,14 @@ public class FormHierarchyActivity extends CollectAbstractActivity implements De
                 && index.getElementMultiplicity() == 0;
     }
 
+    private boolean didDeleteFirstRepeatItem() {
+        return Collect
+                .getInstance()
+                .getFormController()
+                .getFormIndex()
+                .getElementMultiplicity() == 0;
+    }
+
     /**
      * Similar to {@link #goUpLevel}, but makes a less significant step backward.
      * This is only used when the caller knows where to go back to,
@@ -794,7 +802,10 @@ public class FormHierarchyActivity extends CollectAbstractActivity implements De
             //   and you delete an item from the second repeat, it will send you into the
             //   first repeat instead of going back a level as expected.
             goToPreviousEvent();
+        } else if (didDeleteFirstRepeatItem()) {
+            goUpLevel();
         } else {
+            goToPreviousEvent();
             goUpLevel();
         }
     }

